python/pyspark/sql/datasource.py

```diff
@@ -153,7 +191,7 @@ class DataSourceReader(ABC):
     .. versionadded: 4.0.0
     """
 
-    def partitions(self) -> Iterator[Any]:
+    def partitions(self) -> List[InputPartition]:
```

Comment:
I think you should better use `Sequence` instead of `List` so it allows tuple and list.